### PR TITLE
accounts/abi: fix unpacking of arrays of arrays

### PR DESCRIPTION
**Pull request by calmh**
This supersedes #3301 and #3302 while adding a test for both.

### CODE REVIEW

**Comment by calmh** [comment]
This breaks some other unpacking things, let me get back to that :)

**Comment by calmh** [comment]
I think this mishandles dynamic sized slices now, or something like that. However prior to this, statically sized arrays of things other than bytes are instead mishandled. :/ It's not super clear to me what the right solution is as I just started looking on this codebase today... In other similar things I've seen it's common for the method like `toGoType` to return the number of consumed bytes, instead of the caller guessing.

File: accounts/abi/abi.go [hunk 1]
```diff
@@ -274,12 +274,21 @@ func (abi ABI) Unpack(v interface{}, name string, output []byte) error {
 		// struct will match named return values to the struct's field
 		// names
 		case reflect.Struct:
+			idx := 0
```
Comment (VoR0220) [inline accounts/abi/abi.go]: This cannot be what you use to reference the toGoType, as its needed to move through the bytecode.

Comment (calmh) [inline accounts/abi/abi.go]: Sorry, I'm not sure what you mean. I declare idx as an alternative to i, as the index into output must be separate from the index into the number of outputs.

Comment (calmh) [inline accounts/abi/abi.go]: Sorry, I'm not sure what you mean. I declare idx as an alternative to i, as the index into output must be separate from the index into the number of outputs.

Comment (VoR0220) [inline accounts/abi/abi.go]: You need i. Having an index going through the output as 0 throughout all of the outputs breaks the unpacking.

Comment (calmh) [inline accounts/abi/abi.go]: It is not zero. It tracks `i` precisely, except for when it is incremented *more* than `i` for array outputs where `i` is incorrect.

File: accounts/abi/abi.go [hunk 2]
```diff
@@ -274,12 +274,21 @@ func (abi ABI) Unpack(v interface{}, name string, output []byte) error {
...
 			for i := 0; i < len(method.Outputs); i++ {
-				marshalledValue, err := toGoType(i, method.Outputs[i], output)
+				marshalledValue, err := toGoType(idx, method.Outputs[i], output)
 				if err != nil {
 					return err
 				}
 				reflectValue := reflect.ValueOf(marshalledValue)
+				kind := reflectValue.Type().Kind()
```
Comment (obscuren) [inline accounts/abi/abi.go]: This could be `typ := reflectValue.Type()` and you won't have to call `Type` multiple times here, and in the `if` clause.

**Comment by calmh** [comment]
Hi. I'm unfortunately not going to drive this all the way home, as I don't understand all the issues involved. The patch fixes the cases required for my usage and will live as a local fork on my side for the moment. Perhaps someone else with a deeper involvement in the codebase will fix it more generally in the future.

**Comment by obscuren** [comment]
There's no need to close the PR though. If you've no time to further fix the issue it will likely be done by one of us in due time. Would be good to get your commits in regardless.

**Comment by slothbag** [comment]
Any progress made on this issue?  Would be great to get back on master go-ethereum?

**Comment by VoR0220** [comment]
Going to solve this in a separate PR. Closing for now. May reopen and fashion it with capabilities for dynamic slices based on timelines.

**Pull request closed without merging**